src/hotspot/cpu/x86/stubGenerator_x86_64_string.cpp

```diff
@@ -0,0 +1,2012 @@
+  assert((int) StrIntrinsicNode::LL < 4, "Enum out of range");
+  assert((int) StrIntrinsicNode::UL < 4, "Enum out of range");
+  assert((int) StrIntrinsicNode::UU < 4, "Enum out of range");
+  generate_string_indexof_stubs(fnptrs, StrIntrinsicNode::LL);
+  generate_string_indexof_stubs(fnptrs, StrIntrinsicNode::UL);
+  generate_string_indexof_stubs(fnptrs, StrIntrinsicNode::UU);
+  assert(fnptrs[StrIntrinsicNode::LL] != nullptr, "LL not generated.");
+  assert(fnptrs[StrIntrinsicNode::UL] != nullptr, "UL not generated.");
+  assert(fnptrs[StrIntrinsicNode::UU] != nullptr, "UU not generated.");
+}
+
+void StubGenerator::generate_string_indexof_stubs(address *fnptrs, StrIntrinsicNode::ArgEncoding ae) {
+  StubCodeMark mark(this, "StubRoutines", "stringIndexOf");
+  bool isLL = (ae == StrIntrinsicNode::LL);
+  bool isUL = (ae == StrIntrinsicNode::UL);
+  bool isUU = (ae == StrIntrinsicNode::UU);
+  bool isU = isUL || isUU;  // At least one is UTF-16
+  assert(isLL || isUL || isUU, "Encoding not recognized");
+
+  ////////////////////////////////////////////////////////////////////////////////////////
+  ////////////////////////////////////////////////////////////////////////////////////////
+  //                         AVX2 code
+  ////////////////////////////////////////////////////////////////////////////////////////
+  ////////////////////////////////////////////////////////////////////////////////////////
+  if (VM_Version::supports_avx2()) {  // AVX2 version
+
+    ////////////////////////////////////////////////////////////////////////////////////////
+    ////////////////////////////////////////////////////////////////////////////////////////
+    //                         Code generation explanation:
+    //
+    //  The generator will generate code for three cases:
+    //  1. Both needle and haystack are Latin-1 (single byte) encoded (LL)
+    //  2. Both the needle and haystack are UTF-16 encoded (two bytes per character) (UU)
+    //  3. The haystack is UTF-16 encoded and the needle is Latin-1 encoded (UL)
+    //
+    //  The case of the haystack being Latin-1 and the needle being UTF-16 is short-circuited
+    //  so that we never get called in this case.
+    //
+    //  For the UL case (haystack UTF-16 and needle Latin-1), the needle will be expanded
+    //  onto the stack (for size <= MAX_NEEDLE_LEN_TO_EXPAND) and the UU code will do the work.
+    //  For UL where the needle size is > MAX_NEEDLE_LEN_TO_EXPAND, we default to a
+    //  byte-by-byte comparison (this will be rare).
+    //
+    //  Note that the code assumes MAX_NEEDLE_LEN_TO_EXPAND is >= 32.
+    //
+    //  The UU and LL cases are identical except for the loop increments and loading
+    //  of the characters into registers.  UU loads and compares words, LL - bytes.
+    ////////////////////////////////////////////////////////////////////////////////////////
+    ////////////////////////////////////////////////////////////////////////////////////////
+
```

Comment:
This comment can go right before the method start. Also good to add in the comment the native function parameters.